### ISSUE TITLE
allow overriding ccsid

### ISSUE THREAD

**Issue by msillence** [IBM/JTOpen#68]
original issue https://sourceforge.net/p/jt400/bugs/474/  - updated here

We have the not entirely unique problem of needing to read/write to a table with the wrong ccsid https://stackoverflow.com/questions/37409928/db2-on-as400-encoding-issue-with-jt400

however we also need to write to the database and though getBytes works for getting Strings out ignoring the encoding setBytes does not and is rejected as the wrong type

Also this means we need to replace every getString/setString with getBytes and the conversion

I note that AS400JDBCConnectionImpl has a getConverter(int ccsid) and we could enhance this to allow an overridden ccsid
```
@Override
public ConvTable getConverter(int ccsid) throws SQLException {
  if (ccsid == 0 || ccsid == 1 || ccsid == 65535 || ccsid == -1) {
	  return converter_;
  }
  if (this.forcedCcsid != null) {
	  return super.getConverter(this.forcedCcsid);
  }
  return super.getConverter(ccsid);
}
```

We can configure it with
```
    @Override
public void setProperties(JDDataSourceURL dataSourceUrl, JDProperties properties,
        AS400 as400, Properties info) throws SQLException {
    super.setProperties(dataSourceUrl, properties, as400, info);
    String p = info.getProperty(FORCED_CCSID);
    if (p != null) {
        forcedCcsid = Integer.parseInt(p);
    }
}
```

I know we should fix our tables but as it's consistently wrong in thousands of tables and all applications that use the tables the reality is we are more likely to replace the mainframe than fix the problem.

We can easilly override the existing implementation and have our local fix but injecting it is harder as the AS400JDBCDriver has the implementation hard coded in the private method prepareConnection so we also need to copy the entirity of AS400JDBCDriver

Lastly we need to make sure the original driver isn't registered
```
        while (true) {
            Driver remove = DriverManager.getDriver("jdbc:as400://host");
            DriverManager.deregisterDriver(remove);
        }
```
I fully expect this to be rejected as a proposal but thought it might be useful to others

full code here: https://github.com/jhc-systems/debezium-connector-ibmi/tree/main/jt400-override-ccsid

It seems my suggestion isn't complete and has some undesired results for using the metadata and retrieving the column names

**Comment by ThePrez** [comment]
@msillence, is this something you still would like help with? I can try to queue it up but I know this has been opened for a while, so I want to verify that there's still a need.

**Comment by msillence** [comment]
> @msillence, is this something you still would like help with? I can try to queue it up but I know this has been opened for a while, so I want to verify that there's still a need.

yes, sorry I missed the email about this but we'd very much like to be able to directly map the the ccsid used rather than the ccsid configured

**Comment by msillence** [comment]
I should also add that it turns out that the simple force ccsid doesn't cut it.
It worked for us but testing for our client they had the ccsid applied to the table names and they were being corrupted. I'm not quite sure what's different betwen our systems.
The solution we went for was to have a from and to ccsid so we only change the ccsid when it matches the one we know is wrong.

this is essentially the change:

https://github.com/jhc-systems/debezium-connector-ibmi/blob/main/jt400-override-ccsid/src/main/java/com/ibm/as400/access/AS400JDBCConnectionForcedCcsid.java

The rest of the files are just to make it use the overridden file.

For someone that knows the code better, they might find a better solution

**Comment by bonyjohn** [comment]
@ThePrez @msillence
We’re experiencing a similar issue. We were handling it manually using AS400Text or by casting directly in SQL. It would be fantastic if this could be prioritized and added to the queue.
